test/gtest/uct/test_mem.cc

```diff
@@ -106,25 +108,40 @@ UCS_TEST_P(test_mem, md_alloc) {
         status = uct_md_query(md, &md_attr);
         ASSERT_UCS_OK(status);
 
-        for (nonblock = 0; nonblock <= 1; ++nonblock) {
-            int flags = nonblock ? UCT_MD_MEM_FLAG_NONBLOCK : 0;
+        ucs_for_each_bit(mem_type, md_attr.cap.alloc_mem_types) {
+            params.mem_type = (ucs_memory_type_t)mem_type;
+            for (nonblock = 0; nonblock <= 1; ++nonblock) {
+                if ((nonblock == 1) && (mem_type != UCS_MEMORY_TYPE_HOST)) {
+                    continue;
+                }
+                if ((mem_type == UCS_MEMORY_TYPE_CUDA)
+                    || (mem_type == UCS_MEMORY_TYPE_CUDA_MANAGED)) {
+                    if (!mem_type_ctx_created) {
+                        address = mem_buffer::allocate(1, (ucs_memory_type_t)mem_type);
+                        mem_buffer::release(address, (ucs_memory_type_t)mem_type);
+                        mem_type_ctx_created = 1;
+
+                    }
```

Comment:
hacky
let's just select cuda context in some global place in gtest

Reply:
@yosefe Tried setting context in test_base::init() because this seems to be called as part of most init calls. This does end up in context not set error for cuda alloc tests though. Is it the case that new threads are spawned during different tests?

GPU context is associated with thread that calls SetDevice (or explicitly calls cuCreateCtx) and is available to just thread. However, UCT cuda requires calling thread to have a cuda context associated with it. Say for calling uct_md_malloc. How do we ensure that every thread that every gets run for gtests purposes has a context associated with it? Is there a general thread init function?

Reply:
@yosefe @bureddy I've moved intialization to a test_base::init() now. Can you please take a look?

It turned out that for cuMemAlloc calls, I needed to also retain context when user calls cudaSetDevice(0). Without this, the alloc calls were failing and this was the original source of the above issue and not different threads without a cuda context making cuda driver calls.